Refactor input_xbox_one node

steph-lheureux opened the pull request:
This cleans up the controller input code from last year to be more readable and less redundant. Also added control for roll and a reset button.

### Significant changes
- Some of the code for handling "sticky buttons" was redundant. This version abstracts a what were multiple variables to handle the button/feature states and redundant code to a dictionary of `StickyButton` objects. The `StickyButton` class handles initializing a `StickyButton`'s internal tracking variables, checking and setting the feature state with a function, and a reset function.
- When testing the code for the buttons, they appeared to be unreliable. This was determined by @mike-matera  to be due to button bouncing. Debounce code was added to the `StickyButton` to eliminate this issue
- The left and right buttons (`LB` and `RB`) on the controller now control roll
- The `x-box` button resets all input settings to the default state

### General code cleanup
- Added python-style doc strings and lots of comments
- Changed code styling to match official python guidelines
- Removed unused imports
- Changed `controller` dictionary keys to be the functionality name rather than the button name to make code more readable
- Changed class name to match the node name 
- Changed names of topics to be more descriptive 
- Changed claw message type to be a Bool rather than an array (because at the moment we only have one claw)
- Changed the button which controls "bambi mode" to be the `B` button (B like bambi I guess)

Commit:
Begin refactoring input from xbox (first version)
- Created prev button states dictionary to keep track of the previous state if each button in a dictionary rather than many ungrouped variables
- Created layout dictionary specfying the layout of the controller to make swapping layouts easier
- Cleaned up implementation of bambi mode to reduce redundancy
- Removed a lot of code which will be slowly added back in

Commit:
Eliminated redunant dictionary
Changed the controller mapping such that it maps from the joy button/axes values directly to a name discribing functionality. Button names are indicated with comments
Also changed the name of the node to match the file

Commit:
Added functionality for sticky buttons

Commit:
Added depth triggers

Commit:
Configured A button to be sticky and publish state as claw message

Commit:
Began implementation of __throttle_curve function

Commit:
Moved sticky button code to a class

Commit:
Removed unnessicary tracker of prev button state in sticky button class

Commit:
Added button debouncing

Commit:
Added reset button

Commit:
Added roll fuctionality

Commit:
Removed throttle curve stuff (it was bad)

Commit:
Cleanup code
- Add comments
- Remove useless variables
- Fix whitespace between sections of code

Commit:
Changed topic names to be more descriptive

Commit:
Changed spacing to make things aligned and corrected spellng

liamgilligan committed:
Fixed error in comment and changed hex representation to binary

liamgilligan committed:
Changed roll control for conciseness and functionality

Removed unnecessary logic, changed assignment so that when pilot presses down both bumpers the robot won't roll

liamgilligan committed:
Moved Bambi Mode divison to an if statement for readability

liamgilligan committed:
Fixed capitlization in comment

liamgilligan commented on src/seahawk/seahawk_deck/input_xbox_one.py:
```diff
@@ -21,164 +21,171 @@
+        self.__track_state = (self.__track_state << 1 | cur_button_state) & 0b1111
+
+        # Account for bounce by making sure the last four recorded states
+        # appear to represent a button press. If so, update the feature state
```
Elaborate on what "appear to represent a button press" means. Otherwise content looks good.

steph-lheureux replied:
Comment updated in the [most recent commit](https://github.com/CabrilloRoboticsClub/cabrillo_rov_2023/commit/e7efd045c0572b6d1198cc24afde0a474b85617f) to be more descriptive

liamgilligan commented on src/seahawk/seahawk_deck/input_xbox_one.py:
```diff
@@ -21,164 +21,171 @@
+            self.__feature_state = not self.__feature_state
+        return self.__feature_state
+
+    def reset(self):
```
No references to this function. Evaluate necessity.

liamgilligan replied:
Github didn't show any calls to that functionally initially for some reason, it's being used so conversation is no longer relevant.

liamgilligan commented on src/seahawk/seahawk_deck/input_xbox_one.py:
Again, content looks good. Only thing to keep in mind is that the `claw_control` topic is publishing a `Bool` now instead of an `Int8MultiArray`, meaning we are going to have to modify the claw code at a later date.

That shouldn't prevent a merge, however, since the claw design is not finalized this code isn't breaking anything.

steph-lheureux replied:
Should we make this an issue so we do not forget to update the topic names?

liamgilligan replied:
Sounds good to me, link the issue in this conversation and that should be all.

steph-lheureux replied:
Issue has been issued #183

Commit:
Updated comment as suggested in PR review
"Elaborate on what "appear to represent a button press" means. Otherwise content looks good."
New comment is more specific as what a button press is

mike-matera reviewed (approved):
Looks good to me!
